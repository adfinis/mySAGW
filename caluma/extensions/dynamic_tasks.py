from caluma.caluma_workflow.dynamic_tasks import BaseDynamicTasks, register_dynamic_task
from caluma.caluma_workflow.models import WorkItem


def set_one_workitem_ready(query):
    is_ready = False
    for wi in query:
        wi.status = WorkItem.STATUS_CANCELED
        if not is_ready:
            wi.status = WorkItem.STATUS_READY
            is_ready = True
        wi.save()


class CustomDynamicTasks(BaseDynamicTasks):
    @register_dynamic_task("after-review-document")
    def resolve_after_review(self, case, user, prev_work_item, context):
        review_decision = prev_work_item.document.answers.get(
            question_id="review-document-decision",
        )

        if "reject" in review_decision.value:
            return ["revise-document"]
        if "continue" in review_decision.value:
            return ["circulation"]

        return []

    @register_dynamic_task("after-decision-and-credit")
    def resolve_after_decision_and_credit(self, case, user, prev_work_item, context):
        credit_decision = prev_work_item.document.answers.get(
            question_id="decision-and-credit-decision",
        )

        if "additional-data" in credit_decision.value:
            additional_data = case.work_items.filter(
                task__slug="additional-data",
                status=WorkItem.STATUS_REDO,
            )
            if additional_data.exists():
                set_one_workitem_ready(additional_data)
                return ["additional-data-form", "advance-credits"]

            return ["additional-data", "additional-data-form", "advance-credits"]
        if "define-amount" in credit_decision.value:
            define_amount = case.work_items.filter(
                task__slug="define-amount",
                status=WorkItem.STATUS_REDO,
            )
            if define_amount.exists():
                set_one_workitem_ready(define_amount)
                return []
            return ["define-amount"]

        return []

    @register_dynamic_task("after-define-amount")
    def resolve_after_define_amount(self, case, user, prev_work_item, context):
        decision = prev_work_item.document.answers.get(
            question_id="define-amount-decision",
        )

        if "reject" in decision.value:
            additional_data = case.work_items.filter(
                task__slug="additional-data",
                status=WorkItem.STATUS_REDO,
            )
            if additional_data.exists():
                set_one_workitem_ready(additional_data)
            return ["additional-data"]
        if "continue" in decision.value:
            return ["complete-document"]

        return []

    @register_dynamic_task("redo-define-amount")
    def redo_from_define_amount(self, case, user, prev_work_item, context):
        credit_decision = (
            case.work_items.filter(task_id="decision-and-credit")
            .order_by("-created_at")
            .first()
            .document.answers.get(
                question_id="decision-and-credit-decision",
            )
        )

        if "define-amount" in credit_decision.value:
            return ["decision-and-credit"]

        return []
