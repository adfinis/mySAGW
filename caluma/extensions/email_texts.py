SUBJECT_EN = "New information concerning {dossier_nr}"
BODY_EN = """Dear {first_name} {last_name}

The status of the following application/document has changed:

Reference: {dossier_nr}
You can access your document directly via the following link:
{link}

If you have any questions, please feel free to contact us via sagw@sagw.ch, the telephone number +41 31 306 92 50 or at www.sagw.ch/kontakt-mysagw. Please always quote the reference number if you have a question.

Yours sincerely

Your SAGW team

******

This is an automatically generated email, please do not reply to this message. You can find the contact details of SAGW on the following page:
www.sagw.ch/kontakt-mysagw
"""

SUBJECT_DE = "Neue Informationen zu {dossier_nr}"
BODY_DE = """Guten Tag {first_name} {last_name}

Der Status des nachfolgenden Gesuchs/Dokument hat sich geändert:

Referenz: {dossier_nr}
Über den nachfolgenden Link können Sie direkt auf Ihr Dokument zugreifen:
{link}

Sollten Sie Fragen haben, können Sie uns gerne über sagw@sagw.ch, die Telefonnummer +41 31 306 92 50 oder unter www.sagw.ch/kontakt-mysagw kontaktieren. Gerne bitten wir Sie, bei einer Frage immer die Referenznummer anzugeben.

Freundliche Grüsse

Ihr SAGW-Team

******

Dies ist eine automatisch generierte Email, bitte antworten Sie nicht auf diese Nachricht. Die Kontaktangaben der SAGW können Sie der nachfolgenden Seite entnehmen:
www.sagw.ch/kontakt-mysagw
"""

SUBJECT_FR = "Nouvelles informations concernant {dossier_nr}"
BODY_FR = """Bonjour {first_name} {last_name}

Le statut de la demande/du document ci-dessous a changé:

Référence : {dossier_nr}
Cliquez sur le lien ci-dessous pour ouvrir votre document:
{link}

Si vous avez des questions, n'hésitez pas à nous contacter via sagw@sagw.ch, au numéro de téléphone +41 31 306 92 50 ou à l'adresse www.sagw.ch/kontakt-mysagw. Nous vous prions de toujours indiquer le numéro de référence lorsque vous posez une question.

Avec nos meilleures salutations,

Votre équipe de l'ASSH

******

Ceci est un message électronique généré automatiquement, merci de ne pas y répondre. Vous trouverez les coordonnées de l'ASSH sur la page suivante :
www.sagw.ch/kontakt-mysagw
"""


EMAIL_SUBJECTS = {"de": SUBJECT_DE, "fr": SUBJECT_FR, "en": SUBJECT_EN}
EMAIL_BODIES = {"de": BODY_DE, "fr": BODY_FR, "en": BODY_EN}
