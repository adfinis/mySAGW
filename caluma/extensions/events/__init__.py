# Import all modules in this folder so the event receivers are registered
# automatically without having to define each file as caluma event receiver
# module

from . import case, form, work_item  # noqa: F401
