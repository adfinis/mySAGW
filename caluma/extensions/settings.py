import os

import environ
from django.conf import settings
from django.utils.translation import gettext_lazy as _

env = environ.Env()


# Mailing
settings.EMAIL_HOST = os.environ.get("EMAIL_HOST", "mail")
settings.EMAIL_PORT = os.environ.get("EMAIL_PORT", 1025)
settings.EMAIL_HOST_USER = os.environ.get(
    "EMAIL_HOST_USER", "noreply@mysagw.adfinis.com"
)
settings.SERVER_EMAIL = os.environ.get("SERVER_EMAIL", settings.EMAIL_HOST_USER)
settings.EMAIL_HOST_PASSWORD = os.environ.get("EMAIL_HOST_PASSWORD", "")
settings.EMAIL_USE_TLS = os.environ.get("EMAIL_USE_TLS", False)
from_name = os.environ.get("MAILING_FROM_NAME", "mySAGW")
from_mail = os.environ.get("MAILING_FROM_MAIL", "noreply@mysagw.adfinis.com")
settings.MAILING = {"from_email": from_mail, "from_name": from_name}
settings.MAILING_SENDER = f"{from_name} <{from_mail}>"

# API
settings.API_BASE_URI = os.environ.get("API_BASE_URI", "https://mysagw.local/api/v1")
settings.SELF_URI = os.environ.get("SELF_URI", "https://mysagw.local")
raw_verify_ssl = os.environ.get("API_VERIFY_SSL", "true")
settings.API_VERIFY_SSL = True if raw_verify_ssl == "true" else False

settings.OIDC_ADMIN_CLIENT_ID = os.environ.get("OIDC_ADMIN_CLIENT_ID", "test_client")
settings.OIDC_ADMIN_CLIENT_SECRET = os.environ.get(
    "OIDC_ADMIN_CLIENT_SECRET", "fb13e564-75dd-4fc3-a993-3dad9064e71e"
)
settings.OIDC_TOKEN_ENDPOINT = os.environ.get(
    "OIDC_TOKEN_ENDPOINT",
    "https://mysagw.local/auth/realms/mysagw/protocol/openid-connect/token",
)


# Internationalization
# https://docs.djangoproject.com/en/3.2/topics/i18n/
settings.LANGUAGE_CODE = env.str("LANGUAGE_CODE", "de")
settings.LANGUAGES = [
    ("de", _("German")),
    ("en", _("English")),
    ("fr", _("French")),
]
settings.LOCALIZED_FIELDS_FALLBACKS = {
    "de": ["en", "fr"],
    "fr": ["en", "de"],
    "en": ["de", "fr"],
}
settings.LOCALIZED_FIELDS_EXPERIMENTAL = False

# Case

settings.CASE_STATUS = {
    "submit-document": "submit",
    "review-document": "audit",
    "revise-document": "revise",
    "additional-data": "submit-receipts",
    "define-amount": "decision",
    "complete-document": "decision",
}

settings.APPLICANT_TASK_SLUGS = [
    "submit-document",
    "revise-document",
    "additional-data",
    "additional-data-form",
]

settings.CIRCULATION_TASK_SLUGS = [
    "circulation-decision",
]

settings.REVISION_QUESTIONS = {
    "review-document": [
        "priorisierung-der-antrage-kommentar",
        "review-document-decision",
    ],
    "decision-and-credit": [
        "gesprochener-rahmenkredit",
        "decision-and-credit-remark",
        "decision-and-credit-decision",
    ],
    "define-amount": [
        "define-amount-amount-float",
        "define-amount-remark",
        "define-amount-decision",
    ],
}

settings.ADDITIONAL_DATA_FORM = {
    "periodika-antrag": "periodika-abrechnung",
}

settings.INTERNAL_APPLICATION_FORM_SLUG = "intern"
settings.INTERNAL_APPLICATION_TYPE_QUESTION_SLUG = "intern-gesuchsart"
settings.INTERNAL_APPLICATION_PERIODICS_CHOICES = (
    "intern-gesuchsart-intern-abrechnung-periodika",
    "intern-gesuchsart-intern-antrag-periodika",
)

settings.CASE_ID_CACHE_SECONDS = int(os.environ.get("CASE_ID_CACHE_SECONDS", "60"))

# Validation

settings.BIRTHDATE_SLUG_PART = "geburtsdatum"

# Logging

settings.LOGGING = {
    "version": 1,
    "disable_existing_loggers": False,
    "formatters": {
        "verbose": {
            "format": "{levelname} {asctime} {module} {process:d} {thread:d} {message}",
            "datefmt": "%Y-%m-%d %H:%M:%S",
            "style": "{",
        },
        "simple": {
            "format": "{levelname} {message}",
            "style": "{",
        },
    },
    "filters": {
        "require_debug_true": {
            "()": "django.utils.log.RequireDebugTrue",
        },
        "require_debug_false": {
            "()": "django.utils.log.RequireDebugFalse",
        },
    },
    "handlers": {
        "console": {
            "level": "INFO",
            "filters": ["require_debug_true"],
            "class": "logging.StreamHandler",
            "formatter": "simple",
        },
        "mail_admins": {
            "level": "ERROR",
            "filters": ["require_debug_false"],
            "class": "django.utils.log.AdminEmailHandler",
            "formatter": "verbose",
        },
    },
    "loggers": {
        "django": {
            "handlers": ["console", "mail_admins"],
            "level": "WARNING" if settings.DEBUG else "ERROR",
            "propagate": True,
        },
        "django.request": {
            "handlers": ["console", "mail_admins"],
            "level": "WARNING" if settings.DEBUG else "ERROR",
            "propagate": False,
        },
        "graphql": {
            "handlers": ["console", "mail_admins"],
            "level": "WARNING" if settings.DEBUG else "ERROR",
            "propagate": True,
        },
        "caluma": {
            "handlers": ["console", "mail_admins"],
            "level": "WARNING" if settings.DEBUG else "ERROR",
            "propagate": True,
        },
    },
}
