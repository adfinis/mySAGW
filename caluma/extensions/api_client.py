from datetime import datetime, timedelta

import requests
from oauthlib.oauth2 import BackendApplicationClient
from requests_oauthlib import OAuth2Session

from .settings import settings


class APIClient:
    def __init__(
        self,
        base_uri=settings.API_BASE_URI,
        token=None,
    ):
        self.base_uri = base_uri
        self.token = token

    def get_admin_token(self):
        """
        If needed fetch a (new) token from the oidc provider.

        The threshold for fetching a new token is 1 minute before expiration.
        :return: dict
        """
        thresh = datetime.now() + timedelta(minutes=1)

        if self.token is None or self.token["expires_at_dt"] <= thresh:
            client = BackendApplicationClient(client_id=settings.OIDC_ADMIN_CLIENT_ID)
            oauth = OAuth2Session(client=client)
            token = oauth.fetch_token(
                token_url=settings.OIDC_TOKEN_ENDPOINT,
                client_id=settings.OIDC_ADMIN_CLIENT_ID,
                client_secret=settings.OIDC_ADMIN_CLIENT_SECRET,
                verify=settings.API_VERIFY_SSL,
                scope="openid",
            )
            token["expires_at_dt"] = datetime.utcfromtimestamp(int(token["expires_at"]))

            self.token = token["access_token"]

        return self.token

    def get(self, url, *args, **kwargs):
        return self._request(requests.get, url, *args, **kwargs)

    def post(self, url, *args, **kwargs):
        return self._request(requests.post, url, *args, **kwargs)

    def _request(self, method, url, *args, **kwargs):
        token = kwargs.pop("token", self.token)
        headers = {
            "Authorization": f"Bearer {token}",
            "Content-Type": "application/vnd.api+json",
        }
        resp = method(f"{self.base_uri}{url}", *args, **kwargs, headers=headers)
        resp.raise_for_status()
        return resp.json()
