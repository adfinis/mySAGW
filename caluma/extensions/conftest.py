import datetime
from pathlib import Path

import pytest
from django.core.management import call_command

from caluma.caluma_form.models import Answer, Form
from caluma.caluma_user.models import BaseUser
from caluma.caluma_workflow.api import complete_work_item, skip_work_item, start_case
from caluma.caluma_workflow.models import Workflow

from . import api_client
from .settings import settings


@pytest.fixture
def caluma_data(db):
    call_command("loaddata", Path.cwd() / "caluma" / "data" / "form-config.json")
    call_command("loaddata", Path.cwd() / "caluma" / "data" / "workflow-config.json")


@pytest.fixture
def user():
    return BaseUser(username="name", claims={"sub": "test"}, token=b"eyToken")


@pytest.fixture
def document_review_case(db, user):
    case = start_case(
        workflow=Workflow.objects.get(pk="document-review"),
        form=Form.objects.get(pk="circulation-form"),
        user=user,
    )

    Answer.objects.create(
        question_id="circulation-decision",
        document=case.document,
        value="circulation-decision-approved",
    )

    return case


@pytest.fixture
def circulation(db, document_review_case, user):
    case = document_review_case
    skip_work_item(case.work_items.get(task_id="submit-document"), user)

    case.work_items.get(task_id="review-document").document.answers.create(
        question_id="review-document-decision",
        value="review-document-decision-continue",
    )

    complete_work_item(case.work_items.get(task_id="review-document"), user)

    return case.work_items.get(task_id="circulation").child_case


@pytest.fixture
def identities_mock(requests_mock):
    data = {
        "data": [
            {
                "type": "identities",
                "id": "4d70c3f9-1e69-4e82-9010-1a28d12b1e58",
                "attributes": {
                    "email": "test@example.com",
                    "first-name": "Winston",
                    "last-name": "Smith",
                    "language": "en",
                },
                "relationships": {
                    "interests": {"meta": {"count": 0}, "data": []},
                    "additional-emails": {"meta": {"count": 0}, "data": []},
                    "phone-numbers": {"meta": {"count": 0}, "data": []},
                    "addresses": {"meta": {"count": 0}, "data": []},
                },
            }
        ]
    }

    return requests_mock.get(f"{settings.API_BASE_URI}/identities", json=data)


@pytest.fixture
def get_token_mock(mocker):
    data = {
        "access_token": "eyToken",
        "expires_in": 300,
        "refresh_expires_in": 0,
        "token_type": "Bearer",
        "not-before-policy": 0,
        "scope": ["profile", "email"],
        "expires_at": 1632468505.0576448,
        "expires_at_dt": datetime.datetime(2021, 9, 24, 7, 28, 25),
    }

    return mocker.patch.object(
        api_client.OAuth2Session,
        "fetch_token",
        return_value=data,
    )


@pytest.fixture
def case_access_create_request_mock(requests_mock):
    data = {
        "data": {
            "type": "case-accesses",
            "id": "ee6f5d24-b351-4b44-9d55-7dbd2c19d16a",
            "attributes": {
                "case-id": "97d001cf-8d07-4733-aa17-542ed83e8582",
                "email": None,
            },
            "relationships": {
                "identity": {
                    "data": {
                        "type": "identities",
                        "id": "d7b118a7-ce53-48b7-9b05-be148f154f14",
                    }
                }
            },
        }
    }

    return requests_mock.post(f"{settings.API_BASE_URI}/case/accesses", json=data)


@pytest.fixture
def case_access_request_mock(requests_mock):
    data = {
        "data": [{"attributes": {"case-id": "994b72cc-6556-46e5-baf9-228457fa309f"}}]
    }

    return requests_mock.get(f"{settings.API_BASE_URI}/case/accesses", json=data)


@pytest.fixture
def case_access_event_mock(
    requests_mock, identities_mock, get_token_mock, case_access_create_request_mock
):
    data = {
        "data": [
            {
                "type": "case-accesses",
                "id": "6b27043f-70be-403d-94d4-0b6684f954a3",
                "attributes": {
                    "case-id": "994b72cc-6556-46e5-baf9-228457fa309f",
                    "email": None,
                },
                "relationships": {
                    "identity": {
                        "data": {
                            "type": "identities",
                            "id": "19c90e69-0398-4ed7-9bde-1ed13627b1f7",
                        }
                    }
                },
            },
            {
                "type": "case-accesses",
                "id": "675c738c-cc3c-4877-97fc-1ac195a8cb2d",
                "attributes": {
                    "case-id": "994b72cc-6556-46e5-baf9-228457fa309f",
                    "email": "test-dont-send@example.com",
                },
                "relationships": {"identity": {"data": None}},
            },
        ],
        "included": [
            {
                "type": "identities",
                "id": "19c90e69-0398-4ed7-9bde-1ed13627b1f7",
                "attributes": {
                    "idp-id": "267796f2-ae48-4235-93d7-bf26e8ba66bb",
                    "first-name": "Winston",
                    "last-name": "Smith",
                    "email": "test-send@example.com",
                    "language": "de",
                },
            },
        ],
    }

    return requests_mock.get(f"{settings.API_BASE_URI}/case/accesses", json=data)
