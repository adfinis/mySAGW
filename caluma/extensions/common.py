from .api_client import APIClient


def get_api_user_attributes(token, idp_id):
    client = APIClient(token=token)
    result = client.get(f"/identities?filter%5BidpIds%5D={idp_id}")
    attributes = result["data"][0]["attributes"]
    return attributes
