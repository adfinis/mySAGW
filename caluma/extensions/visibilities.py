from django.db.models import F, Q

from caluma.caluma_core.types import Node
from caluma.caluma_core.visibilities import BaseVisibility, Union, filter_queryset_for
from caluma.caluma_form import models as form_models
from caluma.caluma_form.schema import Answer, Document, Form, Option, Question
from caluma.caluma_workflow.schema import Case, Flow, Task, Workflow, WorkItem
from caluma.extensions.common import get_cases_for_user
from caluma.extensions.settings import settings


class StaffVisibility(BaseVisibility):
    @filter_queryset_for(Node)
    def filter_queryset_for_all(self, node, queryset, info):
        user = info.context.user
        if "sagw" in user.groups or "admin" in user.groups:
            return queryset

        return queryset.none()


class CreateOrAssignVisibility(BaseVisibility):
    @filter_queryset_for(Node)
    def filter_queryset_for_all(self, node, queryset, info):  # pragma: no cover
        user = info.context.user
        return queryset.filter(created_by_user=user.username)

    @filter_queryset_for(Answer)
    def filter_queryset_for_answer(self, node, queryset, info):
        return queryset.filter(
            Q(
                document__family__in=self.filter_queryset_for_document(
                    None, form_models.Document.objects, info
                )
            )
            | (
                Q(
                    question__slug__in=[
                        question
                        for sublist in settings.REVISION_QUESTIONS.values()
                        for question in sublist
                    ]
                )
                & Q(
                    document__family__in=self.filter_queryset_for_document(
                        None,
                        form_models.Document.objects,
                        info,
                        list(settings.REVISION_QUESTIONS.keys()),
                    )
                )
            )
        )

    @filter_queryset_for(Document)
    def filter_queryset_for_document(
        self, node, queryset, info, tasks=settings.APPLICANT_TASK_SLUGS
    ):
        user = info.context.user
        case_ids = get_cases_for_user(user)

        work_item = Q(family__work_item__task__slug__in=tasks) | (
            Q(family__work_item__isnull=True)
            & Q(family__case__workflow__pk="document-review")
        )

        access_to_case = (
            Q(family__case__family_id__in=case_ids)
            | Q(family__work_item__case__family_id__in=case_ids)
            | Q(created_by_user=user.username)
        )

        floating_row_document = (
            Q(pk=F("family_id"))
            & Q(form__is_published=False)
            & Q(family__case__isnull=True)
            & Q(family__work_item__isnull=True)
            & Q(created_by_user=user.username)
        )

        row_document_on_applicant_form = Q(family__form__is_published=True) | Q(
            family__work_item__task__slug__in=settings.APPLICANT_TASK_SLUGS
        )
        return queryset.filter(
            Q(work_item & access_to_case)
            | floating_row_document
            | Q(row_document_on_applicant_form & access_to_case)
        )

    @filter_queryset_for(WorkItem)
    def filter_queryset_for_workitem(self, node, queryset, info):
        user = info.context.user
        case_ids = get_cases_for_user(user)
        return queryset.filter(
            Q(case__pk__in=case_ids) | Q(created_by_user=user.username),
            task__slug__in=[
                *settings.APPLICANT_TASK_SLUGS,
                *list(settings.REVISION_QUESTIONS.keys()),
            ],
        )

    @filter_queryset_for(Case)
    def filter_queryset_for_case(self, node, queryset, info):
        user = info.context.user
        case_ids = get_cases_for_user(user)
        return queryset.filter(Q(pk__in=case_ids) | Q(created_by_user=user.username))

    @filter_queryset_for(Task)
    @filter_queryset_for(Flow)
    @filter_queryset_for(Workflow)
    @filter_queryset_for(Form)
    @filter_queryset_for(Option)
    @filter_queryset_for(Question)
    def filter_queryset_for_blueprints(self, node, queryset, info):
        return queryset


class MySAGWVisibility(Union):
    visibility_classes = [StaffVisibility, CreateOrAssignVisibility]
