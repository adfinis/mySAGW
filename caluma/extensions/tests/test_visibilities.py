import pytest

from caluma.caluma_analytics import (
    models as analytics_models,
    schema as analytics_schema,
)
from caluma.caluma_form import models as form_models, schema as form_schema
from caluma.caluma_user.models import OIDCUser
from caluma.caluma_workflow import models as workflow_models, schema as workflow_schema
from caluma.caluma_workflow.api import complete_work_item
from caluma.extensions.settings import settings
from caluma.extensions.visibilities import MySAGWVisibility


@pytest.mark.usefixtures("_caluma_data")
@pytest.mark.parametrize(
    "user_name,visibility_map",
    [
        (
            "admin",
            {
                "Form": 8,
                "WorkItem": 16,
                "Document": 10,
                "Case": 6,
                "Question": 22,
                "Answer": 7,
                "AnalyticsTable": 1,
            },
        ),
        (
            "test",
            {
                "Form": 8,
                "WorkItem": 7,
                "Document": 4,
                "Case": 3,
                "Question": 22,
                "Answer": 4,
                "AnalyticsTable": 0,
            },
        ),
    ],
)
def test_visibilities_default(
    db,
    case_access_event_mock,
    create_document_review_case,
    admin_info,
    admin_user,
    requests_mock,
    analytics_table,
    user_name,
    visibility_map,
    work_item_factory,
    snapshot,
):
    admin_info.context.user.username = user_name
    admin_info.context.user.groups = [user_name]
    admin_info.context.user.group = user_name

    admin_user = OIDCUser(
        b"sometoken",
        {"sub": "admin", settings.OIDC_GROUPS_CLAIM: ["admin"]},
    )
    user = OIDCUser(
        b"sometoken",
        {"sub": user_name, settings.OIDC_GROUPS_CLAIM: [user_name]},
    )

    def prepare_case(case_identifier):
        case = create_document_review_case()
        case.meta = {"identifier": case_identifier}
        case.save()
        complete_work_item(case.work_items.get(task_id="submit-document"), admin_user)
        case.work_items.get(task_id="review-document").document.answers.create(
            question_id="review-document-decision",
            value="review-document-decision-continue",
        )
        complete_work_item(case.work_items.get(task_id="review-document"), admin_user)
        return case

    # case_1 with access
    case_1 = prepare_case("access")
    data = {"data": [{"attributes": {"case-id": str(case_1.pk)}}]}
    requests_mock.get(f"{settings.API_BASE_URI}/case/accesses", json=data)

    # case_2 with circulation invite
    case_2 = prepare_case("circulation")
    case_2_circulation = case_2.work_items.get(task_id="circulation").child_case
    work_item_factory(
        case=case_2_circulation,
        assigned_users=[user.username],
        task=workflow_models.Task.objects.get(slug="circulation-decision"),
        document__form=form_models.Form.objects.get(slug="circulation-form"),
        child_case=None,
    )

    # case_3 no access for non-admin
    prepare_case("only admin")

    test_map = [
        (
            form_models.Form,
            form_schema.Form,
            ["pk"],
        ),
        (
            workflow_models.WorkItem,
            workflow_schema.WorkItem,
            ["task_id", "case__workflow_id"],
        ),
        (
            form_models.Document,
            form_schema.Document,
            ["work_item__task_id", "case__workflow_id"],
        ),
        (
            workflow_models.Case,
            workflow_schema.Case,
            ["family__meta__identifier", "workflow__pk"],
        ),
        (
            form_models.Question,
            form_schema.Question,
            ["pk"],
        ),
        (
            form_models.Answer,
            form_schema.Answer,
            ["question__pk", "value"],
        ),
        (
            analytics_models.AnalyticsTable,
            analytics_schema.AnalyticsTable,
            ["pk"],
        ),
    ]

    vis = MySAGWVisibility()

    for model, node, values_list in test_map:
        queryset = model.objects.all()
        filtered = vis.filter_queryset(node, queryset, admin_info)
        assert filtered.count() == visibility_map[model.__name__]
        values_to_snapshot = [
            tuple(xi for xi in v if xi is not None)
            for v in filtered.values_list(*values_list)
        ]

        assert sorted(values_to_snapshot) == snapshot(name=model.__name__)
