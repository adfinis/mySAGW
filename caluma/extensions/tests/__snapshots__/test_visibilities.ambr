# serializer version: 1
# name: test_visibilities_default[admin-visibility_map0][AnalyticsTable]
  list([
    tuple(
      'live-five-term',
    ),
  ])
# ---
# name: test_visibilities_default[admin-visibility_map0][Answer]
  list([
    tuple(
      'review-document-decision',
      'review-document-decision-continue',
    ),
    tuple(
      'review-document-decision',
      'review-document-decision-continue',
    ),
    tuple(
      'review-document-decision',
      'review-document-decision-continue',
    ),
    tuple(
      'suggest-adult-allow',
      'foo',
    ),
    tuple(
      'suggest-adult-allow',
      'foo',
    ),
    tuple(
      'suggest-adult-allow',
      'foo',
    ),
    tuple(
      'verteilplan-nr',
      'verteilplan-nr-vp-2021',
    ),
  ])
# ---
# name: test_visibilities_default[admin-visibility_map0][Case]
  list([
    tuple(
      'access',
      'circulation',
    ),
    tuple(
      'access',
      'document-review',
    ),
    tuple(
      'circulation',
      'circulation',
    ),
    tuple(
      'circulation',
      'document-review',
    ),
    tuple(
      'only admin',
      'circulation',
    ),
    tuple(
      'only admin',
      'document-review',
    ),
  ])
# ---
# name: test_visibilities_default[admin-visibility_map0][Document]
  list([
    tuple(
      'circulation',
    ),
    tuple(
      'circulation',
    ),
    tuple(
      'circulation',
    ),
    tuple(
      'circulation-decision',
    ),
    tuple(
      'document-review',
    ),
    tuple(
      'document-review',
    ),
    tuple(
      'document-review',
    ),
    tuple(
      'review-document',
    ),
    tuple(
      'review-document',
    ),
    tuple(
      'review-document',
    ),
  ])
# ---
# name: test_visibilities_default[admin-visibility_map0][Form]
  list([
    tuple(
      'additional-data-form',
    ),
    tuple(
      'advance-credits-form',
    ),
    tuple(
      'body-perhaps',
    ),
    tuple(
      'circulation-form',
    ),
    tuple(
      'decision-and-credit-form',
    ),
    tuple(
      'define-amount-form',
    ),
    tuple(
      'quittungen',
    ),
    tuple(
      'review-document-form',
    ),
  ])
# ---
# name: test_visibilities_default[admin-visibility_map0][Question]
  list([
    tuple(
      'additional-data-adresse',
    ),
    tuple(
      'additional-data-bank',
    ),
    tuple(
      'additional-data-bank-town',
    ),
    tuple(
      'additional-data-fibu',
    ),
    tuple(
      'additional-data-iban',
    ),
    tuple(
      'additional-data-name',
    ),
    tuple(
      'additional-data-quittungen',
    ),
    tuple(
      'additional-data-row-quittung',
    ),
    tuple(
      'additional-data-text',
    ),
    tuple(
      'advance-credit-amount-float',
    ),
    tuple(
      'circulation-comment',
    ),
    tuple(
      'circulation-decision',
    ),
    tuple(
      'decision-and-credit-decision',
    ),
    tuple(
      'define-amount-amount-float',
    ),
    tuple(
      'define-amount-decision',
    ),
    tuple(
      'dossier-nr',
    ),
    tuple(
      'gesprochener-rahmenkredit',
    ),
    tuple(
      'review-document-decision',
    ),
    tuple(
      'sektion',
    ),
    tuple(
      'submit-button',
    ),
    tuple(
      'suggest-adult-allow',
    ),
    tuple(
      'verteilplan-nr',
    ),
  ])
# ---
# name: test_visibilities_default[admin-visibility_map0][WorkItem]
  list([
    tuple(
      'circulation',
      'document-review',
    ),
    tuple(
      'circulation',
      'document-review',
    ),
    tuple(
      'circulation',
      'document-review',
    ),
    tuple(
      'circulation-decision',
      'circulation',
    ),
    tuple(
      'finish-circulation',
      'circulation',
    ),
    tuple(
      'finish-circulation',
      'circulation',
    ),
    tuple(
      'finish-circulation',
      'circulation',
    ),
    tuple(
      'invite-to-circulation',
      'circulation',
    ),
    tuple(
      'invite-to-circulation',
      'circulation',
    ),
    tuple(
      'invite-to-circulation',
      'circulation',
    ),
    tuple(
      'review-document',
      'document-review',
    ),
    tuple(
      'review-document',
      'document-review',
    ),
    tuple(
      'review-document',
      'document-review',
    ),
    tuple(
      'submit-document',
      'document-review',
    ),
    tuple(
      'submit-document',
      'document-review',
    ),
    tuple(
      'submit-document',
      'document-review',
    ),
  ])
# ---
# name: test_visibilities_default[test-visibility_map1][AnalyticsTable]
  list([
  ])
# ---
# name: test_visibilities_default[test-visibility_map1][Answer]
  list([
    tuple(
      'review-document-decision',
      'review-document-decision-continue',
    ),
    tuple(
      'review-document-decision',
      'review-document-decision-continue',
    ),
    tuple(
      'suggest-adult-allow',
      'foo',
    ),
    tuple(
      'suggest-adult-allow',
      'foo',
    ),
  ])
# ---
# name: test_visibilities_default[test-visibility_map1][Case]
  list([
    tuple(
      'access',
      'document-review',
    ),
    tuple(
      'circulation',
      'circulation',
    ),
    tuple(
      'circulation',
      'document-review',
    ),
  ])
# ---
# name: test_visibilities_default[test-visibility_map1][Document]
  list([
    tuple(
      'circulation-decision',
    ),
    tuple(
      'document-review',
    ),
    tuple(
      'document-review',
    ),
    tuple(
      'review-document',
    ),
  ])
# ---
# name: test_visibilities_default[test-visibility_map1][Form]
  list([
    tuple(
      'additional-data-form',
    ),
    tuple(
      'advance-credits-form',
    ),
    tuple(
      'body-perhaps',
    ),
    tuple(
      'circulation-form',
    ),
    tuple(
      'decision-and-credit-form',
    ),
    tuple(
      'define-amount-form',
    ),
    tuple(
      'quittungen',
    ),
    tuple(
      'review-document-form',
    ),
  ])
# ---
# name: test_visibilities_default[test-visibility_map1][Question]
  list([
    tuple(
      'additional-data-adresse',
    ),
    tuple(
      'additional-data-bank',
    ),
    tuple(
      'additional-data-bank-town',
    ),
    tuple(
      'additional-data-fibu',
    ),
    tuple(
      'additional-data-iban',
    ),
    tuple(
      'additional-data-name',
    ),
    tuple(
      'additional-data-quittungen',
    ),
    tuple(
      'additional-data-row-quittung',
    ),
    tuple(
      'additional-data-text',
    ),
    tuple(
      'advance-credit-amount-float',
    ),
    tuple(
      'circulation-comment',
    ),
    tuple(
      'circulation-decision',
    ),
    tuple(
      'decision-and-credit-decision',
    ),
    tuple(
      'define-amount-amount-float',
    ),
    tuple(
      'define-amount-decision',
    ),
    tuple(
      'dossier-nr',
    ),
    tuple(
      'gesprochener-rahmenkredit',
    ),
    tuple(
      'review-document-decision',
    ),
    tuple(
      'sektion',
    ),
    tuple(
      'submit-button',
    ),
    tuple(
      'suggest-adult-allow',
    ),
    tuple(
      'verteilplan-nr',
    ),
  ])
# ---
# name: test_visibilities_default[test-visibility_map1][WorkItem]
  list([
    tuple(
      'circulation',
      'document-review',
    ),
    tuple(
      'circulation',
      'document-review',
    ),
    tuple(
      'circulation-decision',
      'circulation',
    ),
    tuple(
      'review-document',
      'document-review',
    ),
    tuple(
      'review-document',
      'document-review',
    ),
    tuple(
      'submit-document',
      'document-review',
    ),
    tuple(
      'submit-document',
      'document-review',
    ),
  ])
# ---
